proposals/gc/MVP.md

```diff
@@ -594,112 +594,52 @@ In particular, `ref.null` is typed as before, despite the introduction of `none`
+  - `br_on_cast_fail $l null? ht : [t0* (ref null ht')] -> [t0* (ref null2? ht)]`
     - iff `$l : [t0* t']`
-    - and `t <: anyref`
-    - and `t <: t'`
+    - and `(ref null3? ht) <: t'`
+    - and `ht <: tht` and `ht' <: tht` where `tht` is a common super type
+    - and `null? = null2? =/= null3?`
   - passes operand along with branch, plus possible extra args
+  - if `null?` is present, does not branch on null, otherwise does
 
-* `ref.as_data` converts to a data reference
-  - `ref.as_data : [anyref] -> [(ref data)]`
-  - traps if reference is not compound data
-  - equivalent to `(block $l (param anyref) (result (ref data)) (br_on_data $l) (unreachable))`
+Note: Cast instructions allow an operand of sibling reference type, even though this cannot possibly succeed. That's because subtyping allows to forget type information and go to a common supertype, so by the subtype substitutibility property, it would be accepted in any case. The given typing rules merely allow this type to also propagate to the result, which avoids the need to compute a least upper bound between the operand type and the target type in the typing algorithm. In practice, it is sufficient to test that both types share the same top heap type.
 
-* `ref.as_array` converts to an array reference
-  - `ref.as_array : [anyref] -> [(ref array)]`
-  - traps if reference is not an array
-  - equivalent to `(block $l (param anyref) (result (ref array)) (br_on_array $l) (unreachable))`
+Note: The [reference types](https://github.com/WebAssembly/reference-types) and [typed function references](https://github.com/WebAssembly/function-references)already introduce similar `ref.is_null`, `br_on_null`, and `br_on_non_null` instructions. These can now be interpreted as syntactic sugar:
 
-* `ref.as_i31` converts to an integer reference
-  - `ref.as_i31 : [anyref] -> [(ref i31)]`
-  - traps if reference is not an integer
-  - equivalent to `(block $l (param anyref) (result (ref i31)) (br_on_i31 $l) (unreachable))`
+* `ref.is_null` is equivalent to `ref.test null ht`, where `tht` is the suitable bottom type (`none`, `nofunc`, or `noextern`)
 
-Note: The [reference types](https://github.com/WebAssembly/reference-types) and [typed function references](https://github.com/WebAssembly/function-references)already introduce similar `ref.is_null`, `br_on_null`, and `br_on_non_null` instructions.
-
-Note: The `br_on_*` instructions allow an operand of unrelated reference type, even though this cannot possibly succeed. That's because subtyping allows to forget that information, so by the subtype substitutibility property, it would be accepted in any case. The given typing rules merely allow this type to also propagate to the result, which avoids the need to compute a least upper bound between the operand type and the target type in the typing algorithm.
-
-
-#### Casts
+* `br_on_null` is equivalent to `br_on_cast null ht`, where `tht` is the suitable bottom type
```

Comment:
Should "is the suitable bottom type" here instead be "is the heap type of the operand" (c.f. `ref.as_non_null` below).

Reply:
No, br_on_null effectively is a downcast to null, with the other side of the union being empty.

But I just realised that br_on_(non)_null do not pass on the null value, so they are not exactly equivalent. Tweaked text.